test/base/Pool.test.hh

```diff
@@ -0,0 +1,124 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Pool.test.hh
+//---------------------------------------------------------------------------//
+
+#include "base/Pool.hh"
+
+namespace celeritas_test
+{
+using celeritas::MemSpace;
+using celeritas::Ownership;
+
+//---------------------------------------------------------------------------//
+// MOCK POOLS
+//---------------------------------------------------------------------------//
+
+struct MockElement
+{
+    int    atomic_number = 0;
+    double atomic_mass;
+};
+
+struct MockMaterial
+{
+    double                           number_density;
+    celeritas::PoolRange<MockElement> elements;
+};
+
+template<Ownership W, MemSpace M>
+struct MockParamsPools
+{
+    template<class T>
+    using Pool = celeritas::Pool<T, W, M>;
+
+    celeritas::Pool<MockElement, W, M>  elements;
+    celeritas::Pool<MockMaterial, W, M> materials;
+    int                                 max_element_components;
+
+    //! Assign from another set of pools
+    template<Ownership W2, MemSpace M2>
+    MockParamsPools& operator=(const MockParamsPools<W2, M2>& other)
+    {
+        elements               = other.elements;
+        materials              = other.materials;
+        max_element_components = other.max_element_components;
+        return *this;
+    }
+};
+
+template<Ownership W, MemSpace M>
+struct MockStatePools
+{
+    template<class T>
+    using Pool = celeritas::Pool<T, W, M>;
+
+    celeritas::Pool<int, W, M> matid;
+
+    celeritas::size_type size() const { return matid.size(); }
+
+    //! Assign from another set of pools
+    template<Ownership W2, MemSpace M2>
+    MockStatePools& operator=(MockStatePools<W2, M2>& other)
```

Comment:
If the user tries to make an illegal assignment, the error is (likely to) going to be something like:
`during the instantantion of MockStatePool::operator=, can not find a matching operator= for the Pool<int...> operator= on line 67.  I.e. a long error output.

The alternative would to use `enable_if` (see the nlohmann/json.hpp) but since that line would need to be copy/pasted/adapted for each Pool holder class, it is unclear were the cost of using/seeing those `enable_if` is worth the clarification in the output.

Reply:
Yep, right now I'm trying to sort through all the typical use cases and get them compiling, then I'm going to see if there's a straightforward way to simplify and improve the class interface and error messages.

Reply:
@pcanal I've substantially reduced the code duplication and improved the failure mechanisms -- now you get a nicely worded static assertion failure when you try to do an invalid copy. operation